### PR TITLE
feat: Hashing IDs Currency Pair Strategy

### PR DESCRIPTION
**Pull request by davidterpay**
- implementing a new currency pair strategy that will broadcast (sha256 hash of each CP, raw price) per vote.
- implements the same caching strategy as other strategies so there should be a similar amount of reads as the delta strategy.
- amount of overhead in terms of VE size should be constant but slightly larger than default.

### CODE REVIEW

**Review by beer-1 — APPROVED** [review]
LGTM 

overhead slightly lager, but I think this way is also good. Thanks for update!

File: abci/strategies/currencypair/hash.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		s.previousHeight = height
+	}
+
+	cp, found := s.idCache[id]
+	if found {
+		return cp, nil
+	}
+
+	// if the currency pair is not found in the cache, attempt to retrieve it from
+	// the x/oracle state by populating the cache with all currency pairs. This
```
Comment (Eric-Warehime) [inline abci/strategies/currencypair/hash.go:68]: Is it a problem in terms of CP removal that we never evict entries from this cache?

**Review by Eric-Warehime — APPROVED** [review]

File: abci/strategies/currencypair/hash.go (hunk 1 above)
Comment (davidterpay) [reply abci/strategies/currencypair/hash.go:68]: This should be evicted once per height.

**Review by aljo242 — APPROVED** [review]

File: abci/strategies/currencypair/hash.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	if !found {
+		return slinkytypes.CurrencyPair{}, fmt.Errorf("currency pair with sha256 hashed ID %d not found in x/oracle state", id)
+	}
+
+	return cp, nil
+}
+
+// CurrencyPairToHashID returns the ID of the given currency pair, by taking the
+// sha256 hash of the currency pair.
+func CurrencyPairToHashID(currencyPair string) (uint64, error) { //nolint
```
Comment (g3ndo9kar9) [inline abci/strategies/currencypair/hash.go:90]: We're truncating the hash here? Is that ok?

**Review by g3ndo9kar9 — COMMENTED** [review]
Can we use this with the delta encoding strategy?

**Comment by davidterpay** [comment]
> Can we use this with the delta encoding strategy?

roll ups need the raw price. otherwise they would need something like ICQ which is not part of the design.

File: abci/strategies/currencypair/hash.go (hunk 2 above)
Comment (davidterpay) [reply abci/strategies/currencypair/hash.go:90]: this is how it was implemented in the forked version, so i just modeled that. assuming it functions as desired.

**Review by g3ndo9kar9 — APPROVED** [review]
lgtm

**Pull request merged**